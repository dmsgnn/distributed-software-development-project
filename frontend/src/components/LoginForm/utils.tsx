import { post } from '@/utils/network';

export enum LoginFields {
	EMAIL = 'login-email',
	PASSWORD = 'login-password',
}

export type LoginFormValues = Record<LoginFields, unknown>;

export function loginUser(formData: LoginFormValues) {
	const requestBody = {
		email: formData[LoginFields.EMAIL],
		password: formData[LoginFields.PASSWORD],
	};

	return post('/api/auth/login', requestBody);
}
