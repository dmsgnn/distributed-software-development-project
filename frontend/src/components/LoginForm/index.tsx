import { Button, FormControl, FormLabel, Input, Text, useToast, VStack } from '@chakra-ui/react';
import { useRouter } from 'next/router';
import { useForm } from 'react-hook-form';
import useMutation from 'use-mutation';
import { LoginFields, LoginFormValues, loginUser } from './utils';

export default function LoginForm() {
	const { handleSubmit, register } = useForm<LoginFormValues>();
	const toast = useToast();
	const router = useRouter();

	const [onSubmit] = useMutation(loginUser, {
		onSuccess: () => {
			router.push('/');
		},
		onFailure: ({ error }) => {
			toast({ title: 'Ops! Something went wrong', status: 'error', description: error?.message });
		},
	});

	return (
		<form onSubmit={handleSubmit(onSubmit)}>
			<VStack align="stretch">
				<FormControl isRequired>
					<FormLabel>Email</FormLabel>
					<Input type="email" placeholder="Email" {...register(LoginFields.EMAIL)} />
				</FormControl>

				<FormControl isRequired>
					<FormLabel>Password</FormLabel>
					<Input type="password" placeholder="Password" {...register(LoginFields.PASSWORD)} />
				</FormControl>

				<Button type="submit" colorScheme="blue">
					Login
				</Button>
			</VStack>
		</form>
	);
}
