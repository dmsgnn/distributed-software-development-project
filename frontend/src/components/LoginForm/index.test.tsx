import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import { act, render, screen, waitFor } from 'test-utils';
import LoginForm from './';

const mockPost = jest.fn();
jest.mock('@/utils/network', () => ({ post: (...args) => mockPost(...args) }));

const mockRouterPush = jest.fn();
jest.mock('next/router', () => ({ useRouter: jest.fn(() => ({ push: mockRouterPush })) }));

describe('LoginForm', () => {
	it('should matches snapshot', () => {
		const { asFragment } = render(<LoginForm />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should be accessible', async () => {
		const { container } = render(<LoginForm />);

		let results: Awaited<ReturnType<typeof axe>>;
		await act(async () => {
			results = await axe(container);
		});

		expect(results).toHaveNoViolations();
	});

	it('should create an api call on submit', async () => {
		mockPost.mockResolvedValue({});
		const user = userEvent.setup();
		render(<LoginForm />);

		const submitButton = screen.getByRole('button', { name: 'Login' });
		expect(submitButton).toBeInTheDocument();

		await user.type(screen.getByPlaceholderText('Email'), 'mail@mail.com');
		await user.type(screen.getByPlaceholderText('Password'), 'password');

		await act(async () => {
			await user.click(submitButton);
		});

		waitFor(() => expect(mockPost).toHaveBeenCalled());
	});

	it('should create an api call on submit (reject)', async () => {
		mockPost.mockRejectedValue({});
		const user = userEvent.setup();
		render(<LoginForm />);

		const submitButton = screen.getByRole('button', { name: 'Login' });
		expect(submitButton).toBeInTheDocument();

		await user.type(screen.getByPlaceholderText('Email'), 'mail@mail.com');
		await user.type(screen.getByPlaceholderText('Password'), 'password');

		await act(async () => {
			await user.click(submitButton);
		});

		waitFor(() => expect(mockPost).toHaveBeenCalled());
	});

	afterAll(() => {
		jest.clearAllMocks();
	});
});
