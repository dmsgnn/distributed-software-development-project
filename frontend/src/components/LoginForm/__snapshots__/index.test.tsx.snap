// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="chakra-stack css-o5l3sd"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="field-:r0:"
          id="field-:r0:-label"
        >
          Email
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-0"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-required="true"
          class="chakra-input css-0"
          id="field-:r0:"
          name="login-email"
          placeholder="Email"
          required=""
          type="email"
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="field-:r1:"
          id="field-:r1:-label"
        >
          Password
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-0"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-required="true"
          class="chakra-input css-0"
          id="field-:r1:"
          name="login-password"
          placeholder="Password"
          required=""
          type="password"
        />
      </div>
      <button
        class="chakra-button css-4xx2wk"
        type="submit"
      >
        Login
      </button>
    </div>
  </form>
</DocumentFragment>
`;
